src/Controller/MakefileToJobTree.ts

```diff
@@ -0,0 +1,63 @@
+import { basename, dirname } from 'path'
+import { exec } from 'child_process'
+import { promisify } from 'util'
+import { Job } from '..'
+import { makefileTraceToJobTree } from './MakefileTraceToJobTree'
+import { MakeTracingError } from './MakeTracingError'
+
+/**
+ * Gets a trace from a makefile by running make with the --trace and --dry-run options.
+ *
+ * @param filePath - the path to the Makefile
+ * @returns the trace.
+ */
+async function getMakefileTrace(options: {
+	filePath?: string
+	targets?: string[]
+}): Promise<string> {
+	let command = 'make --trace --dry-run'
+
+	if (options.filePath) {
+		const fileName = basename(options.filePath)
+		const directoryName = dirname(options.filePath)
+
+		command += ` -f "${fileName}" --directory="${directoryName}"`
+	}
+
+	if (options.targets) {
+		command += ` ${options.targets.join(' ')}`
+	}
+
+	let stdout: string
+	let stderr: string
+	try {
+		;({ stdout, stderr } = await promisify(exec)(command))
```

Comment:
Use `execFile` instead of `exec`. That will execute the process directly instead of creating a shell, and avoids all the pitfalls with escaping the commandline arguments since you pass them in as an array instead of string.